packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,286 @@
+    );
+    event SeededWindow(
+        uint256 indexed windowIndex,
+        uint256 amount,
+        uint256 indexed windowStart,
+        uint256 windowEnd,
+        address indexed rewardToken
+    );
+
+    modifier windowNotLocked(uint256 windowIndex) {
+        require(!merkleWindows[windowIndex].locked, "Window distributions locked");
+        _;
+    }
+
+    constructor(address _timerAddress) public Testable(_timerAddress) {}
+
+    /****************************
+     *
+     * Admin functions
+     *
+     ****************************/
+
+    // Set merkle root for the next available window index and seed allocations. Callable by owner of this
+    // contract. Importantly, we assume that the owner of this contract
+    // correctly chooses an amount `totalRewardsDistributed` that is sufficient
+    // to cover all claims within the `merkleRoot`. Otherwise, a race condition
+    // can be created either within a window (last claimants within a window don't get their claim)
+    // or across windows (claimants from window T take rewards from window T+1).
+    function setWindowMerkleRoot(
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        uint256 windowEnd,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        uint256 indexToSeed = lastSeededIndex;
+        lastSeededIndex = indexToSeed.add(1);
+
+        _seedWindow(indexToSeed, totalRewardsDistributed, windowStart, windowEnd, rewardToken, merkleRoot);
+    }
+
+    // When `window.locked` is true, claims are blocked for that window.
+    function setWindowLock(uint256 windowIndex, bool lockValue) external nonReentrant() onlyOwner {
+        merkleWindows[windowIndex].locked = lockValue;
+    }
+
+    // Overwrite merkle root for specified window index. Owner must deposit `totalRewardsDistributed`
+    // into contract.
+    // TODO: Should we require that `windowIndex <= lastSeededIndex` or should this function just be
+    // a generalized `setWindowMerkleRoot` whereby the owner can set a merkle root for any index?
+    function resetWindowMerkleRoot(
+        uint256 windowIndex,
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        uint256 windowEnd,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        _seedWindow(windowIndex, totalRewardsDistributed, windowStart, windowEnd, rewardToken, merkleRoot);
+    }
+
+    // Emergency methods to transfer rewards into and out of the contract
+    // incase the contract was configured improperly.
+    function withdrawRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransfer(msg.sender, amount);
+    }
+
+    function depositRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
these probably should have events

Reply:
Good catch

Reply:
In addition to the `Transfer` event that will be emitted?

Reply:
Done